src/test/java/harness/utils/ExternalBQDatasets.java

```diff
@@ -1,17 +1,25 @@
 package harness.utils;
 
+import bio.terra.cloudres.google.bigquery.BigQueryCow;
+import com.google.api.services.bigquery.model.Dataset;
+import com.google.api.services.bigquery.model.DatasetReference;
 import com.google.auth.oauth2.GoogleCredentials;
-import com.google.cloud.bigquery.Acl;
 import com.google.cloud.bigquery.BigQuery;
 import com.google.cloud.bigquery.BigQueryOptions;
-import com.google.cloud.bigquery.Dataset;
-import com.google.cloud.bigquery.DatasetInfo;
+import harness.CRLJanitor;
 import harness.TestExternalResources;
 import java.io.IOException;
-import java.util.ArrayList;
+import java.security.GeneralSecurityException;
+import java.util.List;
 import java.util.UUID;
 
-/** Utility methods for creating external BQ datasets for testing workspace references. */
+/**
+ * Utility methods for creating external BQ datasets for testing workspace references. Most methods
+ * in the class use the CRL wrapper around the BQ client library to manipulate external datasets.
+ * This class also includes a method to build the unwrapped BQ client object, which is what we would
+ * expect users to call. Setup/cleanup of external datasets should use the CRL wrapper. Fetching
+ * information from the cloud as a user may do should use the unwrapped client object.
```

Comment:
nit: I don't see what we expect users to use has any effect on what client objects we should use. CRL isn't going to do anything unusual in its GCP calls. In these tests, the reason to use CRL is for a) janitor resource clean up and b) (not for BQ where the cloud library is better) but conveniences over the raw gcp api.services libraries. Let's make it clear when it's important to use CRL wrappers. (it doesn't actually do anything special for deletion)
I think there's a consistency argument to be made for only picking a single type of client object, but if it's much more painful to use the cRL api services, then that's fine, but let's comment that as the reasoning.

ditto other comments.

Reply:
My reason to use CRL in these tests was strictly a) janitor resource clean up. Can you say more about b) the conveniences? For BQ and GCS, the cloud libraries seem good to me in the limited calls we're using it for here anyway.

In general, I think tests should mimic user behavior as best as possible. I don't expect CRL to make any difference here, but it seems unlikely CLI users would use it. That's the only reason I left it out of everything except setup/cleanup.

That's helpful to know about deletion! So for cleanup purposes, the only call we need is for create?

Reply:
1. Things like [polling support](https://github.com/DataBiosphere/terra-cli/blob/60bb9b54659a3f0376262f6a58a7b200d5b62860/src/main/java/bio/terra/cli/service/GoogleAiNotebooks.java#L50-L53) & [value classes for nicer APIs](https://github.com/DataBiosphere/terra-cli/blob/60bb9b54659a3f0376262f6a58a7b200d5b62860/src/main/java/bio/terra/cli/service/GoogleAiNotebooks.java#L24). The cloud libraries are reasonably convenient, but when they don't exist, the raw api services libraries aren't great. We could do more than we're doing now in CRL, but if we're going to do anything to improve the usability, we should do it in a central place i.e. CRL.
2. CRL does some consistent logging forms, which might be helpful for CLI debugging without having to rewrite gcp logging help.
3. Yep, create is the only thing. Janitor needs to handle deleting resources whether or not they exist because we can't guarantee that a resource will be created successfully, or that we'll always publish a creation event if we try to wait until after a resource is created.

Reply:
Gotcha, thanks for the additional information and code links. I think we want all the benefits of CRL in both source code and test infrastructure (e.g. create/delete/set iam policy for buckets/datasets), so I'm going to leave the code that uses CRL for things other than just create, even though it's not required for Janitor.

I slightly prefer using the exact same code users would in the actual tests. In general, I think tests should call code the same way users would -- that was one of my main reasons behind the whole Test Runner effort (e.g. use the client library as a caller would, instead of Spring's integration testing support). Even if it very likely won't make much difference in this case, I think this separation between test and source code is important. Less importantly, I don't think we ever want to be in the situation where we need to add functionality to CRL just to support tests, since that seems an unnecessary overhead.